### PR TITLE
Attention Logit Noise: targeted slice routing regularization (σ=0.05)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

PR #2254 (backbone hidden noise, σ=0.01) failed because additive noise on hidden states compounds through residual connections — too blunt, destabilizes training across all blocks simultaneously. The autopsy identified **attention logit noise** as the mechanistically correct target.

The slice attention logits — the `attn_logits` tensor computed just before the slice-token softmax in `Physics_Attention_Irregular_Mesh` — are the ideal injection point:
- Perturbation affects **slice routing stochasticity** without propagating through residual connections
- This is directly analogous to DropAttention (Zehui Lin et al., 2019), shown to be a strong OOD regularizer by preventing over-reliance on specific attention patterns
- For our CFD model, tandem OOD failures stem from the model routing aft-foil wake nodes to "wrong" slices — logit noise forces robust attention patterns

**Mechanism:** During training only (not eval/EMA), add Gaussian noise N(0, σ²) to the pre-softmax attention logit matrix (`attn_logits`) in all 3 TransolverBlocks. The noise is applied to the cross-slice cosine similarity product, NOT to the per-node slice assignment weights (`slice_weights`). The slice assignment routing remains deterministic.

**Key distinction from PR #2254 (failed):**
- #2254 added noise to `hidden_states` after residual connection → propagates, compounds, destabilizes
- This adds noise to `attn_logits` (pre-softmax, no residual) → contained, targeted, stochastic routing only
- σ=0.05 on logits that are O(1) scale → ~5% routing perturbation, much lower effective noise

### 1. Add CLI flag

In `Config` dataclass:
```python
attn_logit_noise_std: float = 0.0  # std dev of Gaussian noise added to slice attn logits during training
```

### 2. Locate the attention logit computation

In `Physics_Attention_Irregular_Mesh.forward`, find where `attn_logits` (or equivalent: the output of `q @ k.transpose()` scaled by 1/sqrt(d)) is computed, just before the softmax/temperature scaling in the cross-slice attention. This is the tensor of shape `[B, n_head, slice_num, slice_num]` (or `[B, slice_num, slice_num]`).

### 3. Add noise injection

Immediately after computing `attn_logits` and **before** the softmax:
```python
# Attention logit noise — applied during training only, not eval/EMA
if self.training and attn_logit_noise_std > 0:
    attn_logits = attn_logits + torch.randn_like(attn_logits) * attn_logit_noise_std
```

Where `attn_logit_noise_std` is passed in from the config or stored as a module attribute during `__init__`.

**Critical:** Apply ONLY to the **cross-slice attention logits** (the `[B, slice_num, slice_num]` computation), NOT to the per-node-to-slice assignment logits (`slice_weights`). The slice assignment should remain deterministic.

Apply to all 3 TransolverBlocks (pass the noise std when constructing each block).

### 4. Run configuration (2 seeds)

```bash
# Seed 42
cd cfd_tandemfoil && python train.py --agent alphonse \
  --wandb_name "alphonse/attn-logit-noise-s42" \
  --wandb_group "attn-logit-noise" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature \
  --attn_logit_noise_std 0.05 \
  --seed 42

# Seed 73: same flags with --seed 73 --wandb_name "alphonse/attn-logit-noise-s73"
```

Only change vs baseline: `--attn_logit_noise_std 0.05`.

### 5. What to report

Table: p_in, p_oodc, p_tan, p_re for each seed and 2-seed avg vs baseline. W&B run IDs for both seeds.
Key signals: p_oodc and p_tan (OOD generalization metrics).

If results are promising: suggest trying σ=0.1 (stronger routing stochasticity) as a follow-up.

## Baseline (PR #2213, 2-seed avg)

| Metric | Value | Target to beat |
|--------|-------|----------------|
| p_in | 11.979 | < 11.98 |
| p_oodc | 7.643 | < 7.65 |
| p_tan | 28.341 | < 28.34 |
| p_re | 6.300 | < 6.30 |

Baseline W&B runs: `hgml7i2r` (s42), `qic03vrg` (s73)

**Reproduce baseline:**
```bash
cd cfd_tandemfoil && python train.py --agent alphonse --wandb_name "alphonse/baseline-wake-deficit" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT alphonse:

## Results

### Attention Logit Noise σ=0.05 (Cross-Slice Only)

**Configuration:** Added Gaussian noise N(0, 0.05²) to cross-slice attention logits (pre-softmax, shape [B, H, S, S]) during training in all 3 TransolverBlocks. Applied after `attn_logits = q_norm @ k_norm.T * attn_scale` and before `F.softmax`. Per-node-to-slice assignment logits (`slice_weights`) untouched.

| Metric | Seed 42 | Seed 73 | 2-seed avg | Baseline | Target | Verdict |
|--------|---------|---------|------------|----------|--------|---------|
| p_in | 12.3 | 12.4 | **12.36** | 11.979 | < 11.98 | ❌ +3.2% worse |
| p_oodc | 7.6 | 7.5 | **7.55** | 7.643 | < 7.65 | ✅ -1.2% better |
| p_tan | 28.6 | 28.2 | **28.40** | 28.341 | < 28.34 | ❌ +0.2% (noise) |
| p_re | 6.5 | 6.4 | **6.45** | 6.300 | < 6.30 | ❌ +2.4% worse |

**Precise p_in (W&B summary):** seed 42 = 12.2842, seed 73 = 12.43477

- **W&B runs:** lozjp7jd (seed 42), bcgmj504 (seed 73)
- **W&B group:** attn-logit-noise
- **Epochs completed:** 147 (s42), 148 (s73) — both hit 180-min wall-clock timeout
- **val/loss:** 0.3786 (s42), 0.3782 (s73)
- **Peak GPU memory:** ~56 GiB (s42), ~75 GiB (s73)

### Command used
```bash
cd cfd_tandemfoil && CUDA_VISIBLE_DEVICES=0 python train.py --agent alphonse \
  --wandb_name "alphonse/attn-logit-noise-s42" \
  --wandb_group "attn-logit-noise" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature \
  --attn_logit_noise_std 0.05 \
  --seed 42
```

### Implementation notes

The attention logits in the standard (non-learned-kernel) path are:
`attn_logits = q_norm @ k_norm.T * attn_scale`

where `attn_scale` is a learned parameter initialized to 10.0 (can grow during training). Cosine similarity is in [-1, 1], so logits are on the order of [-10, 10]. At σ=0.05, the noise is ~0.5% of the typical logit range — quite small. This is applied in both the learned-kernel branch and the standard normalized-cosine branch.

### What happened

**Negative result — failed to improve over baseline.** The experiment hurt p_in (+3.2%) and p_re (+2.4%), with only a marginal improvement on p_oodc (-1.2%). p_tan is essentially flat.

The poor performance on p_in (in-distribution pressure) is the most concerning signal. This suggests the noise is disrupting fine-grained slice routing that the model has learned for well-posed in-distribution samples, without providing enough OOD benefit to compensate.

Analysis of why this may have failed:
1. **Noise scale vs. logit scale mismatch:** The `attn_scale` is a learnable parameter starting at 10.0. Even at σ=0.05, the relative noise per unit logit is ~0.5% — not enough to meaningfully perturb routing, but apparently enough to add gradient noise during training. This is a dose problem: too small to regularize, large enough to add noise to the loss landscape.
2. **The slice routing is already well-regularized.** The slice_residual_scale, PCGrad, and tandem_ramp already encourage robust representations. Adding attention logit noise on top may be redundant perturbation.
3. **The hypothesis that tandem failures come from incorrect slice routing** may not be accurate — it may be more about the input features than the attention routing.

p_oodc's slight improvement is interesting but within noise range (7.55 vs 7.643). Seed variance on p_in (12.28 vs 12.43) is also concerning — the noise may be making training unstable.

### Suggested follow-ups
- Try much lower σ (0.01 or 0.02) — current σ=0.05 may be in a regime that's neither regularizing nor clearly destabilizing, but adds gradient noise
- Alternatively, apply noise ONLY to tandem samples (gate on `is_tandem_batch`) to target the OOD failure mode without touching in-distribution routing
- The p_oodc slight improvement suggests attention noise may help for purely spatial OOD — could try with σ=0.02 targeted at OOD slices

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2263 because Negative result — failed to beat baseline on any meaningful metric. 2-seed avg: p_in=12.36 (+3.9% vs 11.891 baseline), p_tan=28.40 (+1.0%), p_re=6.45 (+1.3%). Only p_oodc improved marginally (7.55 vs 7.561, essentially noise). The attention logit noise disrupted fine-grained in-distribution slice routing (p_in badly hurt) without providing enough OOD benefit. The hypothesis that tandem failures are caused by incorrect slice routing is likely not the primary driver — the PCGrad + tandem_ramp already regularize gradient conflicts. The noise scale problem: attn_scale is learnable starting at 10.0, so σ=0.05 is ~0.5% relative noise — too small to meaningfully regularize but enough to add gradient noise. Closing as a clear dead end. Thank you alphonse — the mechanistic analysis of why it failed is valuable context..
